reframe/frontend/executors/policies.py

```diff
@@ -55,6 +59,21 @@ def runcase(self, case):
         except BaseException:
             task.fail(sys.exc_info())
         finally:
+
+            # Execute cleanup of dependencies if all dependent cases have been
+            # executed
+            for dep in self.dependency_tree[case]:
+                self.dependency_count[dep] -= 1
+                if self.dependency_count[dep] == 0:
+                    # Check if dep has failed before cleaning
+                    for t in self.stats.tasks():
+                        if t.testcase == dep and t.failed is False:
+                            dependency_clean_up_task = RegressionTask(dep)
+                            dependency_clean_up_task.cleanup(
+                                not self.keep_stage_files,
+                                False)
+                            break
```

Comment:
I need to be sure that the cleanup step is always executed, also in case of exceptions.

Reply:
Why it should be that? If a test fails, its cleanup phase is not executed by just throwing the exception. Also I think that this part might be simplified if you make the serial policy also a `TaskEventListener` so that it gets notified when a task finishes. Then you can adjust the reference count in a cleaner way.